lib/ain-grpc/src/rpc/eth.rs

```diff
@@ -984,160 +983,139 @@ impl MetachainRPCServer for MetachainRPCModule {
-            .map(|block_number| {
-                self.handler
-                    .logs
-                    .get_logs(&input.address, &input.topics, block_number)
-                    .map_err(to_custom_err)
-            })
-            .collect::<RpcResult<Vec<Vec<_>>>>()?
-            .into_iter()
-            .flatten()
-            .map(LogResult::from)
-            .collect::<Vec<_>>();
-        Ok(logs_result)
+        } else {
+            None
+        };
+        let topics = if let Some(topics) = input.topics {
+            match topics {
+                LogRequestTopics::VecOfHashes(inputs) => {
+                    Some(inputs.into_iter().map(|input| vec![input]).collect())
+                }
+                LogRequestTopics::VecOfHashVecs(inputs) => Some(inputs),
+            }
+        } else {
+            None
+        };
+        let curr_block = self.get_block(Some(BlockNumber::Latest))?.header.number;
+        let mut criteria = FilterCriteria {
+            block_hash: input.block_hash,
+            from_block,
+            to_block,
+            addresses: input.address,
+            topics,
+        };
+        criteria
+            .verify_criteria(curr_block)
+            .map_err(RPCError::EvmError)?;
+        let logs = self
+            .handler
+            .filters
+            .get_logs_from_filter(&criteria)
+            .map_err(RPCError::EvmError)?;
+        Ok(logs.into_iter().map(|log| log.into()).collect())
     }
 
     fn new_filter(&self, input: NewFilterRequest) -> RpcResult<U256> {
-        let from_block_number = self.get_block(input.from_block)?.header.number;
-        let to_block_number = self.get_block(input.to_block)?.header.number;
-        if from_block_number > to_block_number {
-            return Err(RPCError::FromBlockGreaterThanToBlock.into());
-        }
+        let from_block = if input.from_block.is_some() {
+            if let Some(BlockNumber::Num(block_num)) = input.from_block {
+                // Allow future block number to be specified
+                Some(U256::from(block_num))
+            } else {
+                Some(self.get_block(input.from_block)?.header.number)
+            }
+        } else {
+            None
+        };
+        let to_block = if input.to_block.is_some() {
+            if let Some(BlockNumber::Num(block_num)) = input.to_block {
+                // Allow future block number to be specified
+                Some(U256::from(block_num))
+            } else {
+                Some(self.get_block(input.to_block)?.header.number)
+            }
+        } else {
+            None
+        };
+        let topics = if let Some(topics) = input.topics {
+            match topics {
+                LogRequestTopics::VecOfHashes(inputs) => {
+                    Some(inputs.into_iter().map(|input| vec![input]).collect())
+                }
+                LogRequestTopics::VecOfHashVecs(inputs) => Some(inputs),
+            }
+        } else {
```

Comment:
Same as above

Reply:
Refactored topics pipeline to use map in [e7ae914](https://github.com/DeFiCh/ain/commit/e7ae914e0564533816ef18415a3266cd5fe2b413)